common/math/math.go

```diff
@@ -34,3 +46,195 @@ func RoundFloat(x float64, prec int) float64 {
+		superMean = append(superMean, result)
+		combined += result
+	}
+	avg := combined / (float64(len(superMean)) - 1)
+	return math.Sqrt(avg), nil
+}
+
+// GeometricAverage is an average which indicates the central tendency or
+// typical value of a set of numbers by using the product of their values
+// The geometric average can only process positive numbers
+func GeometricAverage(values []float64) (float64, error) {
+	if len(values) == 0 {
+		return 0, errZeroValue
+	}
+	product := 1.0
+	for i := range values {
+		if values[i] <= 0 {
+			// cannot use negative or zero values in geometric calculation
+			return 0, errGeometricNegative
+		}
+		product *= values[i]
+	}
+	geometricPower := math.Pow(product, 1/float64(len(values)))
+	return geometricPower, nil
+}
+
+// FinancialGeometricAverage is a modified geometric average to assess
+// the negative returns of investments. It accepts It adds +1 to each
+// This does impact the final figures as it is modifying values
+// It is still ultimately calculating a geometric average
+// which should only be compared to other financial geometric averages
+func FinancialGeometricAverage(values []float64) (float64, error) {
+	if len(values) == 0 {
+		return 0, errZeroValue
+	}
+	product := 1.0
+	for i := range values {
+		if values[i] < -1 {
+			// cannot lose more than 100%, figures are incorrect
+			// losing exactly 100% will return a 0 value, but is not an error
+			return 0, errNegativeValueOutOfRange
+		}
+		// as we cannot have negative or zero value geometric numbers
+		// adding a 1 to the percentage movements allows for differentiation between
+		// negative numbers (eg -0.1 translates to 0.9) and positive numbers (eg 0.1 becomes 1.1)
+		modVal := values[i] + 1
+		product *= modVal
+	}
+	geometricPower := math.Pow(product, 1/float64(len(values)))
+	if geometricPower > 0 {
+		// we minus 1 because we manipulated the values to be non-zero/negative
+		geometricPower--
+	}
+	return geometricPower, nil
+}
+
+// ArithmeticAverage is the basic form of calculating an average.
+// Divide the sum of all values by the length of values
+func ArithmeticAverage(values []float64) (float64, error) {
+	if len(values) == 0 {
+		return 0, errZeroValue
+	}
+	var sumOfValues float64
+	for x := range values {
+		sumOfValues += values[x]
+	}
+	return sumOfValues / float64(len(values)), nil
+}
+
+// SortinoRatio returns sortino ratio of backtest compared to risk-free
+func SortinoRatio(movementPerCandle []float64, riskFreeRate, average float64) (float64, error) {
+	if len(movementPerCandle) == 0 {
+		return 0, errZeroValue
+	}
+	totalNegativeResultsSquared := 0.0
+	for x := range movementPerCandle {
+		if movementPerCandle[x]-riskFreeRate < 0 {
+			totalNegativeResultsSquared += math.Pow(movementPerCandle[x]-riskFreeRate, 2)
+		}
+	}
+	averageDownsideDeviation := math.Sqrt(totalNegativeResultsSquared / float64(len(movementPerCandle)))
+	return (average - riskFreeRate) / averageDownsideDeviation, nil
```

Comment:
riskFreeRate here is being calculated as: relativelyRiskFree := first.Holdings.RiskFreeRate / intervalsPerYear.
Take for example 1 candle size is 1day, therefore intervalsPerYear = 365. But if your strategy is only run for 3months you would want the riskFreeRate for 3 months which would be calced as: relativelyRiskFree := (number of intervals between first and last date for the strategy)/intervalsPerYear

Reply:
Also the riskFreeRate is correctly used on line 215 where you want to subtract the risk free rate per candle

Reply:
True! I need to account for the time. Wouldn't this suffice?
```
	candles := float64(len(movementPerCandle))
	return average - (riskFreeRate*candles)/averageDownsideDeviation, nil
```